File: PyPy3/백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
import sys
input = sys.stdin.readline

N = int(input())
board = [list(map(int, input().split())) for _ in range(N)]

def rotate_board(cnt, board):
    temp_board = [row[:] for row in board]
    for _ in range(cnt):
        rotated = [[0] * N for _ in range(N)]
        for i in range(N):
            for j in range(N):
                rotated[j][N - 1 - i] = temp_board[i][j]
        temp_board = rotated
        
    return temp_board

def move(dir, board):
    rotated = rotate_board(dir, board)
    temp_board = [[0] * N for _ in range(N)]
    
    for i in range(N):
        nums = []
        for j in range(N - 1, -1, -1):
            if rotated[i][j] != 0:
                nums.append(rotated[i][j])
        
        merged = []
        idx = 0
        while idx < len(nums):
            if idx + 1 < len(nums) and nums[idx] == nums[idx + 1]:
                merged.append(nums[idx] * 2)
                idx += 2
            else:
                merged.append(nums[idx])
                idx += 1
        
        while len(merged) < N:
            merged.append(0)
            
        temp_board[i] = merged
            
    return temp_board

def dfs(board, cnt, val):
    if (cnt == 5):
        return val
    
    dir = [0, 1, 2, 3]
    for d in dir:
        temp_board = [row[:] for row in board]
        temp_board = move(d, temp_board)
        cur_max = max(map(max, temp_board))
        val = dfs(temp_board, cnt + 1, max(val, cur_max))
    return val

print(dfs(board, 0, 0))

File: C++17/백준/Gold/12100. 2048 （Easy）/2048 （Easy）.cc
#include <iostream>
#include <vector>
using namespace std;

int N;
int answer;

vector<vector<int>> rotate(vector<vector<int>> board, int dir) {
    vector<vector<int>> temp = board;
    for (int c = 0; c < dir; c++) {
        vector<vector<int>> rotated(N, vector<int>(N, 0));
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                rotated[N - 1 - j][i] = temp[i][j];
            }
        }
        temp = rotated;
    }
    return temp;
}

vector<vector<int>> move(vector<vector<int>> board, int dir) {
    vector<vector<int>> rotated = rotate(board, dir);
    vector<vector<int>> temp(N, vector<int>(N, 0));

    for (int i = 0; i < N; i++) {
        vector<int> nums;
        for (int j = 0; j < N; j++) {
            if (rotated[i][j] != 0) nums.push_back(rotated[i][j]);
        }
        vector<int> merged;
        int idx = 0;
        while (idx < nums.size()) {
            if (idx + 1 < nums.size() && nums[idx] == nums[idx + 1]) {
                merged.push_back(nums[idx] * 2);
                idx += 2;
            }
            else {
                merged.push_back(nums[idx]);
                idx += 1;
            }
        }

        while (merged.size() < N) {
            merged.push_back(0);
        }

        temp[i] = merged;
    }

    return rotate(temp, 4 - dir);
}

void dfs(vector<vector<int>> board, int cnt) {
    if (cnt == 5) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (board[i][j] > answer) answer = board[i][j];
            }
        }
        return;
    }

    vector<int> dir = { 0, 1, 2, 3 };
    for (int i = 0; i < 4; i++) {
        vector<vector<int>> temp_board = move(board, dir[i]);
        dfs(temp_board, cnt + 1);
    }
}

int main() {
    cin >> N;

    vector<vector<int>> board(N, vector<int>(N, 0));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int val;
            cin >> val;
            board[i][j] = val;
        }
    }

    dfs(board, 0);

    cout << answer;
}